sklearn/feature_selection/mutual_info.py

```diff
@@ -0,0 +1,375 @@
+    if x_categorical and y_categorical:
+        return mutual_info_score(x, y)
+    elif x_categorical and not y_categorical:
+        return _compute_mi_cd(y, x, n_neighbors)
+    elif not x_categorical and y_categorical:
+        return _compute_mi_cd(x, y, n_neighbors)
+    else:
+        return _compute_mi_cc(x, y, n_neighbors)
+
+
+def _iterate_csc_columns(X, columns=None):
+    """This little hack returns a densified columns when iterating over a
+    sparse matrix, instead of constructing a sparse matrix for every column
+    that is expensive.
+
+    Parameters
+    ----------
+    X : csc_matrix, shape (n_samples, n_features)
+        Matrix over which to iterate.
+    columns : iterable or None, default None
+        Indices of columns to iterate over. If None, iterate over all columns.
+
+    Yields
+    ------
+    xi : ndarray, shape (n_samples,)
+        Columns of `X` in dense format.
+    """
+    n_features = X.shape[1]
+    X_indices = X.indices
+    X_data = X.data
+    X_indptr = X.indptr
+
+    if columns is None:
+        columns = range(n_features)
+
+    for i in columns:
+        column = np.zeros(X.shape[0])
+        startptr, endptr = X_indptr[i], X_indptr[i + 1]
+        nonzero_indices = X_indices[startptr:endptr]
+        column[nonzero_indices] = X_data[startptr:endptr]
+        yield column
+
+
+def _iterate_dense_columns(X, columns=None):
+    """Simple convenience iterator over ndarray columns.
+
+    Works the same as _iterate_csc_columns, but for dense ndarray.
+    """
+    if columns is None:
+        columns = range(X.shape[1])
+
+    for i in columns:
+        yield X[:, i]
+
+
+def _get_column(X, i):
+    if issparse(X):
+        x = np.zeros(X.shape[0])
+        startptr, endptr = X.indptr[i], X.indptr[i + 1]
+        x[X.indices[startptr: endptr]] = X.data[startptr: endptr]
+    else:
+        x = X[:, i]
+
+    return x
+
+
+class MutualInfoSelector(BaseEstimator, SelectorMixin):
+    """Select a subset of features based on mutual information criterion.
+
+    This class can either select a subset of features which individually have
+    highest mutual information with the target variable, or select a subset of
+    features jointly having maximum mutual information with the target variable
+    using "min-redundancy max-relevance" algorithm [1]_.
+
+    Read more in :ref:`User Guide <MI>`.
+
+    Parameters
+    ----------
+    n_features_to_select : float or int, default 0.5
+        Number of features to select. The value greater than or equal 1 is
+        interpreted as the absolute number of features to select. The value
+        within (0.0, 1.0) is interpreted as the percentage from the initial
+        number of features (rounded down). Half of the features is selected by
+        default. After the class instance was fitted you can decrease
+        `n_features_to_select` and correctly select a smaller subset of
+        features without refitting.
+
+    use_redundancy : bool, default True
```

Comment:
Indeed `MutualInfoSelector` now implements two algorithms (and the name implies that). But what's wrong about it? The simpler algorithm (without redundancy consideration) is still useful, especially when the number of features is large. I like the ability to control algorithm quality vs. time complexity. To my eyes it fits into scikit-learn's style too.

There is an alternative: we can return `relevance_` and drop this option. Although I don't insist on any variant (including removing this option and not returning `relevance_`).